device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,161 @@
+            synchronized (lock) {
+                if(connectionStatus == ConnectionStatus.CONNECTING) {
+                    try {
+                        log.debug("Opening the device client instance " + getClientId() + " ...");
+                        openClient();
+                        connectionStatus = ConnectionStatus.CONNECTED;
+                        break;
+                    }
+                    catch (Exception ex) {
+                        if (ex.getCause() instanceof TransportException && ((TransportException) ex.getCause()).isRetryable()) {
+                            log.warn("Transport exception thrown while opening DeviceClient instance " + getClientId() + ", retrying: ", ex);
+                        } else {
+                            log.error("Non-retryable exception thrown while opening DeviceClient instance " + getClientId() + ": ", ex);
+                            connectionStatus = ConnectionStatus.DISCONNECTED;
+                            throw ex;
+                        }
+                    }
+                }
+            }
+
+            try {
+                log.debug("Sleeping for 10 secs before attempting another open()");
+                Thread.sleep(SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS * 1000);
+            }
+            catch (InterruptedException ex) {
+                throw new RuntimeException("InterruptedException in thread sleep: ", ex);
+            }
+        }
+    }
+
+    public void closeNow() {
+        synchronized (lock) {
+            try {
+                log.debug("Closing the client instance " + getClientId() + " ...");
+                closeClient();
+            }
+            catch (IOException e) {
+                log.error("Exception thrown while closing client instance " + getClientId() + " : ", e);
+            } finally {
+                connectionStatus = ConnectionStatus.DISCONNECTED;
+            }
+        }
+    }
+
+    public void open() throws IOException {
+        synchronized (lock) {
+            if(connectionStatus == ConnectionStatus.DISCONNECTED) {
+                connectionStatus = ConnectionStatus.CONNECTING;
+            } else {
+                return;
+            }
+        }
+        doConnect();
+    }
+
+    @Override
+    public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext) {
+        Pair<IotHubConnectionStatusChangeCallback, Object> suppliedCallbackPair = this.suppliedConnectionStatusChangeCallback;
+        if (throwable == null)
+        {
+            System.out.println("CONNECTION STATUS UPDATE FOR CLIENT: " + getClientId() + " - " + status + ", " + statusChangeReason);
+        }
+        else
+        {
+            System.out.println("CONNECTION STATUS UPDATE FOR CLIENT: " + getClientId() + " - " + status + ", " + statusChangeReason + ", " + throwable.getMessage());
```

Comment:
Might be worthwhile to just call throwable.printStackTrace instead of appending the throwable.getMessage(). getMessage() does not include the full stack trace, so some information is lost

Reply:
Unlike .NET, Java doesn't have a property on the exception itself that contains the stacktrace, so this is the best way to print it out when using System.out